src/ott/neural/models.py

```diff
@@ -43,137 +46,112 @@ class ICNN(neuraldual.BaseW2NeuralDual):
     )
-    self.w_zs = w_zs
-
-    # positive definite potential (the identity mapping or linear OT)
-    self.pos_def_potential = layers.PosDefPotentials(
-        self.dim_data,
-        num_potentials=1,
-        kernel_init=lambda *_: factor,
-        bias_init=lambda *_: mean,
+
+  def _get_wx(self, dim: int) -> nn.Module:
+    return layers.PosDefPotentials(
+        rank=self.rank,
+        num_potentials=dim,
+        use_linear=True,
         use_bias=True,
+        kernel_diag_init=nn.initializers.zeros,
+        kernel_lr_init=self.init_fn,
+        kernel_linear_init=self.init_fn,
+        # TODO(michalk8): why us this for bias?
+        bias_init=self.init_fn,
     )
 
-    # subsequent layers re-injected into convex functions
-    w_xs = []
-    for i in range(self.num_hidden):
-      w_xs.append(
-          nn.Dense(
-              self.dim_hidden[i],
-              kernel_init=self.init_fn(self.init_std),
-              bias_init=initializers.constant(0.),
-              use_bias=True,
-          )
+  def _get_pos_def_potentials(self) -> layers.PosDefPotentials:
+    kwargs = {
+        "num_potentials": 1,
+        "use_linear": True,
+        "use_bias": True,
+        "bias_init": nn.initializers.zeros
+    }
+
+    if self.gaussian_map_samples is None:
+      return layers.PosDefPotentials(
+          rank=self.rank,
+          kernel_diag_init=nn.initializers.ones,
+          kernel_lr_init=nn.initializers.zeros,
+          kernel_linear_init=nn.initializers.zeros,
+          **kwargs,
       )
-    # final layer, to output number
-    w_xs.append(
-        nn.Dense(
-            1,
-            kernel_init=self.init_fn(self.init_std),
-            bias_init=initializers.constant(0.),
-            use_bias=True,
-        )
-    )
-    self.w_xs = w_xs
-
-  @staticmethod
-  def _compute_gaussian_map_params(
-      samples: Tuple[jnp.ndarray, jnp.ndarray]
-  ) -> Tuple[jnp.ndarray, jnp.ndarray]:
-    from ott.tools.gaussian_mixture import gaussian
-    source, target = samples
-    # print(source)
-    # print(type(source))
-    g_s = gaussian.Gaussian.from_samples(source)
-    g_t = gaussian.Gaussian.from_samples(target)
-    lin_op = g_s.scale.gaussian_map(g_t.scale)
-    b = jnp.squeeze(g_t.loc) - jnp.linalg.solve(lin_op, jnp.squeeze(g_t.loc))
-    lin_op = matrix_square_root.sqrtm_only(lin_op)
-    return jnp.expand_dims(lin_op, 0), jnp.expand_dims(b, 0)
-
-  @staticmethod
-  def _compute_identity_map_params(
-      input_dim: int
-  ) -> Tuple[jnp.ndarray, jnp.ndarray]:
-    A = jnp.eye(input_dim).reshape((1, input_dim, input_dim))
-    b = jnp.zeros((1, input_dim))
-    return A, b
 
-  @nn.compact
-  def __call__(self, x: jnp.ndarray) -> float:  # noqa: D102
-    z = self.act_fn(self.w_xs[0](x))
-    for i in range(self.num_hidden):
-      z = jnp.add(self.w_zs[i](z), self.w_xs[i + 1](x))
-      z = self.act_fn(z)
-    z += self.pos_def_potential(x)
-    return z.squeeze()
+    source, target = self.gaussian_map_samples
```

Comment:
something i am a bit unsure about is, in the current setup, the gaussian quadratic potential (i.e. just one) is added at the very end, regardless of all the rest. Is there a way to ensure its scale is therefore reasonable / meaningful / large enough compared to the rest of the initialization?

Reply:
Good point, will check #90 for answers